File: filosofos_python/ejecutar_procesos.py
"""
Script para ejecutar únicamente la solución con PROCESOS.

Uso:
    python ejecutar_procesos.py [num_filosofos] [duracion_segundos]

Ejemplo:
    python ejecutar_procesos.py 5 30
"""

import sys
import time
import multiprocessing
from solucion_procesos.MesaIPC import MesaIPC
from solucion_procesos.ProcesoFilosofo import ProcesoFilosofo


def main():
    # Parsear argumentos de línea de comandos
    num_filosofos = 5
    duracion = 30
    
    if len(sys.argv) > 1:
        try:
            num_filosofos = int(sys.argv[1])
        except ValueError:
            print("Error: El número de filósofos debe ser un entero.")
            sys.exit(1)
    
    if len(sys.argv) > 2:
        try:
            duracion = int(sys.argv[2])
        except ValueError:
            print("Error: La duración debe ser un entero.")
            sys.exit(1)
    
    print("\n" + "="*70)
    print("PROBLEMA DE LOS FILÓSOFOS COMENSALES - SOLUCIÓN CON PROCESOS")
    print("="*70)
    print(f"Configuración:")
    print(f"  - Número de filósofos: {num_filosofos}")
    print(f"  - Duración: {duracion} segundos")
    print("="*70 + "\n")
    
    mesa_ipc = None
    filosofos = []
    
    try:
        # Crear la mesa IPC
        mesa_ipc = MesaIPC(num_filosofos)
        
        # Crear los filósofos
        filosofos = [ProcesoFilosofo(i, mesa_ipc) for i in range(num_filosofos)]
        
        print("Iniciando procesos...\n")
        
        # Iniciar todos los procesos
        for filosofo in filosofos:
            filosofo.iniciar()
        
        print(f"\nSimulación corriendo por {duracion} segundos...")
        print("Presiona Ctrl+C para detener antes.\n")
        
        # Ejecutar por el tiempo especificado
        time.sleep(duracion)
        
    except KeyboardInterrupt:
        print("\n\nSimulación interrumpida por el usuario.\n")
    
    finally:
        # Terminar todos los procesos
        print("\nTerminando procesos...")
        for filosofo in filosofos:
            filosofo.terminar()
        
        print("\n" + "="*70)
        print("Fin de la simulación")
        print("="*70 + "\n")
        
        # Imprimir estadísticas finales (después de que todos terminaron)
        if mesa_ipc:
            mesa_ipc.imprimir_estadisticas()
        
        # Limpiar recursos IPC
        if mesa_ipc:
            mesa_ipc.limpiar()



if __name__ == "__main__":
    # Necesario en Windows para multiprocessing
    multiprocessing.freeze_support()
    main()

File: filosofos_java/src/EjecutarProcesos.java
import procesos.MesaIPC;
import procesos.ProcesoFilosofo;
import java.util.ArrayList;
import java.util.List;

/**
 * Script para ejecutar la solución con PROCESOS (simulados con threads IPC).
 * 
 * Nota: Java no soporta fork() verdadero como Unix/Linux. Esta implementación
 * simula procesos usando threads con comunicación mediante semáforos y memoria atómica.
 * 
 * Uso:
 *     java EjecutarProcesos [num_filosofos] [duracion_segundos]
 * 
 * Ejemplo:
 *     java EjecutarProcesos 5 30
 */
public class EjecutarProcesos {
    
    public static void main(String[] args) {
        // Parsear argumentos de línea de comandos
        int numFilosofos = 5;
        int duracion = 30;
        
        if (args.length > 0) {
            try {
                numFilosofos = Integer.parseInt(args[0]);
            } catch (NumberFormatException e) {
                System.err.println("Error: El número de filósofos debe ser un entero.");
                System.exit(1);
            }
        }
        
        if (args.length > 1) {
            try {
                duracion = Integer.parseInt(args[1]);
            } catch (NumberFormatException e) {
                System.err.println("Error: La duración debe ser un entero.");
                System.exit(1);
            }
        }
        
        System.out.println("\n" + "=".repeat(70));
        System.out.println("PROBLEMA DE LOS FILÓSOFOS COMENSALES - SOLUCIÓN CON PROCESOS");
        System.out.println("=".repeat(70));
        System.out.println("Configuración:");
        System.out.println("  - Número de filósofos: " + numFilosofos);
        System.out.println("  - Duración: " + duracion + " segundos");
        System.out.println("=".repeat(70) + "\n");
        
        MesaIPC mesaIPC = null;
        List<ProcesoFilosofo> filosofos = new ArrayList<>();
        
        try {
            // Crear la mesa IPC
            mesaIPC = new MesaIPC(numFilosofos);
            
            // Crear los filósofos
            for (int i = 0; i < numFilosofos; i++) {
                filosofos.add(new ProcesoFilosofo(i, mesaIPC));
            }
            
            System.out.println("Iniciando procesos...\n");
            
            // Iniciar todos los procesos
            for (ProcesoFilosofo filosofo : filosofos) {
                filosofo.iniciar();
            }
            
            System.out.println("\nSimulación corriendo por " + duracion + " segundos...");
            System.out.println("Presiona Ctrl+C para detener antes.\n");
            
            // Ejecutar por el tiempo especificado
            Thread.sleep(duracion * 1000L);
            
        } catch (InterruptedException e) {
            System.out.println("\n\nSimulación interrumpida por el usuario.\n");
        } finally {
            // Terminar todos los procesos
            System.out.println("\nTerminando procesos...");
            for (ProcesoFilosofo filosofo : filosofos) {
                filosofo.terminar();
            }
            
            // Dar tiempo para que terminen
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            
            System.out.println("\n" + "=".repeat(70));
            System.out.println("Fin de la simulación");
            System.out.println("=".repeat(70) + "\n");
            
            // Imprimir estadísticas finales (después de que todos terminaron)
            if (mesaIPC != null) {
                mesaIPC.imprimirEstadisticas();
            }
        }
        
        // Dar tiempo para que los threads terminen
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        System.exit(0);
    }
}
